plugins/bigquery/dbt/include/bigquery/macros/materializations/incremental.sql

```diff
@@ -94,54 +137,69 @@
   {%- set partitions = config.get('partitions', none) -%}
   {%- set cluster_by = config.get('cluster_by', none) -%}
 
+  {% set on_schema_change = incremental_validate_on_schema_change(config.get('on_schema_change'), default='ignore') %}
+
   {{ run_hooks(pre_hooks) }}
 
+  {# -- first check whether we want to full refresh for source view or config reasons #}
+  {% set trigger_full_refresh = (full_refresh_mode or existing_relation.is_view) %}
+  {% do log('full refresh mode: %s' % trigger_full_refresh) %}
+
   {% if existing_relation is none %}
       {% set build_sql = create_table_as(False, target_relation, sql) %}
+  
   {% elif existing_relation.is_view %}
       {#-- There's no way to atomically replace a view with a table on BQ --#}
       {{ adapter.drop_relation(existing_relation) }}
       {% set build_sql = create_table_as(False, target_relation, sql) %}
+  
   {% elif full_refresh_mode %}
       {#-- If the partition/cluster config has changed, then we must drop and recreate --#}
       {% if not adapter.is_replaceable(existing_relation, partition_by, cluster_by) %}
           {% do log("Hard refreshing " ~ existing_relation ~ " because it is not replaceable") %}
           {{ adapter.drop_relation(existing_relation) }}
       {% endif %}
       {% set build_sql = create_table_as(False, target_relation, sql) %}
+  
   {% else %}
-     {% set dest_columns = adapter.get_columns_in_relation(existing_relation) %}
-
-     {#-- if partitioned, use BQ scripting to get the range of partition values to be updated --#}
-     {% if strategy == 'insert_overwrite' %}
-
-        {% set missing_partition_msg -%}
-          The 'insert_overwrite' strategy requires the `partition_by` config.
-        {%- endset %}
-        {% if partition_by is none %}
-          {% do exceptions.raise_compiler_error(missing_partition_msg) %}
-        {% endif %}
-
-        {% set build_sql = bq_insert_overwrite(
-            tmp_relation,
-            target_relation,
-            sql,
-            unique_key,
-            partition_by,
-            partitions,
-            dest_columns) %}
-
-     {% else %}
-       {#-- wrap sql in parens to make it a subquery --#}
-       {%- set source_sql -%}
-         (
-           {{sql}}
-         )
-       {%- endset -%}
-
-       {% set build_sql = get_merge_sql(target_relation, source_sql, unique_key, dest_columns) %}
+     {% set tmp_relation = make_temp_relation(target_relation) %}
```

Comment:
**Stylistically,** I think we can reorder the logic below to avoid repetition:
```sql
{% set tmp_relation = make_temp_relation(target_relation) %}
{% if on_schema_change != 'ignore' %}
  {# only create the temp table if it's needed. can we make this a free query? #}
  {% do run_query(create_table_as(True, tmp_relation, sql)) %}
  {% set schema_changed = check_for_schema_changes(tmp_relation, target_relation) %}
  {% if schema_changed %}
    {% do process_schema_changes(on_schema_change, tmp_relation, existing_relation) %}
  {% endif %}
{% endif %}
{% set dest_columns = adapter.get_columns_in_relation(existing_relation) %}
{% set build_sql = dbt_bq_get_incremental_sql(strategy, tmp_relation, target_relation, sql, unique_key, partition_by, partitions, dest_columns) %}
```

**Substantively,** there is something subtle going on here: We're creating a temp table first thing, in order to power schema comparison. I think it's fair to document that the "price" of dynamic schema evolution is the need to produce that temp table, but today, dbt does _not_ produce a temp table for the `merge` strategy, and produces that temp table as part of a script in the `insert_overwrite` strategy.

I'm wondering if we could adjust the temp table creation to be a "free" ($0) query, i.e. by adding a `where false limit 0` condition. So like:
```sql
{% set tmp_sql %}
  select * from ({{ sql }}) where false limit 0
{% endset %}
{% do run_query(create_table_as(True, tmp_relation, tmp_sql)) %}
```

Reply:
Implemented this change on snowflake and rolling to other providers.